proposals/NNNN-noncopyable-generics.md

```diff
@@ -0,0 +1,975 @@
+actors can always conform to `Copyable`, because a reference to an object can
+be copied regardless of what is contained in the object.
+
+<!-- What sets apart a layout protocol is that at runtime, it is possible to query
+whether a type conforms (for more details, see a later section). -->
+
+<!-- TODO: make sure this is exhaustive about what other non-nominal types are
+copyable like metatypes, tuples, element packs -->
+
+### General type parameters
+
+All generic type parameters default to carrying a `Copyable` constraint:
+
+```swift
+struct FileDescriptor: ~Copyable { /* ... */ }
+
+// signature <T where T: Copyable>
+func genericFn<T>(_ t: T) /* where T: Copyable */ {
+  return copy t  // OK
+}
+
+genericFn(FileDescriptor())  // ERROR: FileDescriptor is not Copyable
+genericFn([1, 2]) // OK: Array<Int> is Copyable
+```
+
+In [SE-0390](0390-noncopyable-structs-and-enums.md) the prefix `~` syntax was
+introduced only in the inheritance clause of a nominal type to "suppress" its
+Copyable conformance. This proposal completes the picture by defining the
+semantics of `~Copyable` as the inverse of a constraint that cancels out the
+default requirement or conformance to `Copyable`. The definition of inverses for
+types other than `Copyable` is outside the scope of this proposal.
+
+A generic parameter can have its implicit `Copyable` constraint removed by
+applying its inverse `~Copyable` to the subject as a constraint:
+
+```swift
+// signature <T>
+func identity<T: ~Copyable>(_ t: borrowing T) -> T  {
+  return copy t  // ERROR: 't' is noncopyable
+}
+
+identity(FileDescriptor())  // OK, FileDescriptor is not Copyable
+identity([1, 2, 3])  // OK, even though Array<Int> is Copyable
+```
+
+> **Key Idea:** Applying an inverse `T: ~Copyable` does _not_ prevent a
+> `Copyable` type from being substituted for `T`, i.e., it does _not_ narrow the
+> kinds of types permitted, like a constraint would. This is the reason why
+> `~Copyable` is referred to as the _inverse_ of `Copyable` rather than a
+> constraint itself. When applying an inverse `~C` to a type variable implicitly
+> constrained to `C`, that implicit constraint is nullified. Removing any kind
+> of constraint from a generic parameter _broadens_ the kinds of types
+> permitted.
+
+As with a concrete noncopyable type, any generic type parameter that does not
+conform `Copyable` must use one of the ownership modifiers `borrowing`,
+`consuming`, or `inout`, when it appears as the type of a function's parameter.
+For details on these parameter ownership modifiers, see
+[SE-377](0377-parameter-ownership-modifiers.md).
+
+### Structs and enums
+Struct and enum types default to implicitly conforming to `Copyable`:
+
+```swift
+struct DataSet /* : Copyable */ {
+  var samples: [Double]
+}
+```
+
+As with any other generic parameter, there is a `Copyable` constraint on the
+`Elm` in this `List` too:
+
+```swift
+enum List<Elm> /* : Copyable where Elm: Copyable */ {
```

Comment:
Is there a reason to use `Elm` everywhere instead of `Element`? Do you want to avoid confusion with `Collection.Element`?

Reply:
(a reason other than "it's shorter :)

Reply:
No good reason